### ISSUE TITLE
replace GNU make with something else

### ISSUE THREAD

**Issue by boneskull** [mochajs/mocha#2352]
[scripty](https://npmjs.com/package/scripty) is a thing and so is [p-s](https://npmjs.com/package/p-s).

I don't want the overhead of a full-blown task runner like Grunt or Gulp, as I don't want to rely on plugins.

Thoughts?

@kentcdodds Have you seen scripty?  p-s looks similar.  Maybe you guys should pool efforts? 

cc @searls
cc @mochajs/core

**Comment by kentcdodds** [comment]
Yeah, I've seen scripty. Pretty different approaches to the problem. I'm bias, but I feel like p-s is better. scripty requires you add a ton of folders and files to your project which is to much work for a simple script IMO. p-s is basically: package.json scripts in a JavaScript file (meaning you get dynamic logic and comments).

**Comment by kentcdodds** [comment]
Yeah. Give p-s a try. You'll like it.

**Comment by searls** [comment]
A couple scripty notes:

* if your motivation is simply making your package.json smaller, then p-s should be fine, but you're still writing shell scripts as strings in a JavaScript file, which can be awkward when you have many scripts of decent complexity
* scripty's intention is primarily to let Shell scripts be Shell scripts, rather than trying to cram them inside of JavaScript strings; npm is limited to `sh`, as well, whereas scripty is glad to spawn any shebangable scripting language
* the Windows script dichotomy is something I'm not sure how scripty can solve. Most teams I'm on are unix-only, and perhaps it's the only situation under which I'd use scripty I'm definitely open to suggestions if there's anything that can be done reliably short of "wait for Windows to run bash".

tl;dr scripty is a tool for writing a project's scripts in first class files (with syntax highlighting, multiple lines, the whole works) with some added understanding of hierarchical grouping of tasks.

**Comment by kentcdodds** [comment]
Notes in response to @searls notes 😃

> but you're still writing shell scripts as strings in a JavaScript file, which can be awkward when you have many scripts of decent complexity

True. And if things get complicated, you can always invoke a stout file yourself (can even be a JavaScript file). Just like with normal npm scripts

> Most teams I'm on are unix-only

That may be true for some teams. But many teams are multiplatform and even more so is open source. p-s has a solid cross-platform story.

> scripty is a tool for writing a project's scripts in first class files

That's great and I think it solves that problem really well. My issue is when I'm just calling out to another binary like ESLint or Webpack or AVA, I don't want to have an entire file for a single line to invoke the CLI. I like that p-s gives me the flexibility to both call out to CLIs like this and also really simply break out to a file when necessary.

**Comment by TimothyGu** [comment]
Is there a reason why we need to replace GNU Make right now?

Also have you considered [ShellJS](http://documentup.com/shelljs/shelljs)?

**Comment by boneskull** [comment]
> Is there a reason why we need to replace GNU Make right now?

Who said it has to happen right now?  😝 

In general, I would prefer to only use Node.js-based tools to maximize portability.

ShellJS is cool--`shx` is even more applicable to our use case--but it doesn't really solve the same problem `make` does.

**Comment by kentcdodds** [comment]
If you wanted to use `shx`, you'd still want a way to organize/document your scripts. `p-s` can do that for you and it works great with the ShellJS binary :+1: (Though honestly, I've never felt the need to use it).

**Comment by boneskull** [comment]
I have [a branch](/boneskull/mocha/tree/v3.0.0-ps) which starts to leverage `p-s`.  I wanted to ensure the `mocha.js` bundle was built upon `npm version`, and that the bundle had a banner w/ its version number.

It's kind of weird to have both `make` and `p-s` at the same time.  It'd be more developer-friendly if we migrated all at once.  

At any rate, if @mochajs/core has any opinion on `p-s` or another tool, I'm all ears.

**Comment by ScottFreeCode** [comment]
This should come up when I search GitHub issues for "Makefile".

I've mentioned this on Gitter, but I have code prepared that just converts the Makefile targets to NPM scripts and it turned out not that bad. I plan on opening a PR (at which point the whole team can weigh in and evaluate it) after the following are merged as they touch some of the same code:
- #2922 
- #2918 
- #2886 
- #2900 

Now that CI is working again, I am going to try rebasing these as applicable to confirm that the tests pass so we can start merging them. With any luck it'll be fairly clear-cut to finish reviewing any that still need review.
